### PR TITLE
Virgo Startup, Pid file, and more options

### PR DESCRIPTION
**Pull request by rjemanuele**

### CODE REVIEW

File: lib/agent.lua [hunk 1]
```diff
@@ -94,15 +102,35 @@ function Agent:start(options)
...
-      --err, msg = virgo.write_pid(options.pidFile)
-      --if err then
-      --  pcall(function()
-      --    local pid = fs.readFileSync(options.pidFile)
-      --    logging.error(fmt('Agent in use (pid: %d, path: %s)', pid, options.pidFile))
-      --    process:exit(3)
-      --  end)
-      --end
-      callback()
+      fs.open(options.pidFile, 'a+', function(err, fd)
```
Comment (rphillips) [inline lib/agent.lua]: 'w' should truncate the pidFile.

Comment (rjemanuele) [inline lib/agent.lua]: 'w' does truncate the log file... regardless of if you have the lock or not.  that's why many programs use both a lock file (contents are usually empty) and a separate pid file.

File: agent.lua [hunk 2]
```diff
@@ -93,15 +101,33 @@ function Agent:start(options)
...
+          local errno = ffi.errno();
+          local errstr = ffi.C.strerror(errno)
+          logging.error(fmt('Agent lock file flock error (path: %s): %s', options.lockFile, ffi.string(errstr)))
+          ffi.C.free(errstr)
+          fs.close(fd)
+          local pid
+          pcall(function()
+            pid = fs.readFileSync(options.pidFile)
+          end)
+          logging.error(fmt('Agent in use (pid: %s, path: %s)', tostring(pid), options.pidFile))
```
Comment (rphillips) [inline agent.lua]: i think pid can be used here as uninitialized if the pcall fails.

**Comment by rphillips** [comment]
otherwise, +1

**Comment by rphillips** [comment]
i think we move the file locking code into virgo-base

**Pull request merged**